### PR TITLE
Defer

### PR DESCRIPTION
**Pull request by lemaitre-aneo**

### CODE REVIEW

File: csharp/src/Defer.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+using System.Diagnostics;
+
+using JetBrains.Annotations;
+
+namespace ArmoniK.Utils;
+
+/// <summary>
+///   Wraps an action that will be called when the object is disposed
+/// </summary>
+public class Defer : IDisposable
```
Comment (wkirschenmann) [inline csharp/src/Defer.cs:27]: https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces

Comment (wkirschenmann) [reply csharp/src/Defer.cs:27]: `DeferManager` or `DeferHandler`

File: csharp/src/Defer.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  public Defer(Action deferred)
+    => deferred_ = deferred;
+
+  private Defer()
+    => deferred_ = null;
+
+  /// <inheritdoc />
+  public void Dispose()
+  {
+    var deferred = deferred_;
```
Comment (wkirschenmann) [inline csharp/src/Defer.cs:51]: Not very found of the approach. You maximize the chance of concurrent call to dispose and do nothing about it... In release, you do not even detect it. IMHO, the risk is too high of having a double dispose.

Comment (wkirschenmann) [reply csharp/src/Defer.cs:51]: Suggestion : https://learn.microsoft.com/en-us/dotnet/api/system.threading.interlocked.exchange?view=netstandard-2.0

File: csharp/src/Defer.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    if (deferred is null)
+    {
+      return;
+    }
+
+    deferred.Invoke();
```
Comment (wkirschenmann) [inline csharp/src/Defer.cs:57]: In the rest of ArmoniK, usually, we'd write `deferred_()` but it's the same.

File: csharp/src/Defer.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    // Check for race condition on Dispose
+    Debug.Assert(deferred_ is not null);
+
+    deferred_ = null;
+    GC.SuppressFinalize(this);
```
Comment (wkirschenmann) [inline csharp/src/Defer.cs:63]: Not sure we want to risk doing this many times. 
Hence `Interlocked.Exchange`

File: csharp/src/Defer.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+  /// <summary>
+  ///   Constructs a Disposable object that calls a specific action when disposed
+  /// </summary>
+  /// <param name="deferred">Action to be called at Dispose</param>
+  [PublicAPI]
+  public Defer(Action deferred)
+    => deferred_ = deferred;
+
+  private Defer()
```
Comment (wkirschenmann) [inline csharp/src/Defer.cs:45]: Is this required? Why not simply use `Defer(null)` line 41? Or `Defer(Action deferred = null)` line 50?

Comment (lemaitre-aneo) [reply csharp/src/Defer.cs]: I'm not sure we want the public interface to allow a null action.

File: csharp/src/Deferrer.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  ~Deferrer()
+    => Dispose();
+
+  /// <summary>
+  ///   Constructs a Disposable object that calls a specific action when disposed
+  /// </summary>
+  /// <param name="deferred">Action to be called at Dispose time</param>
+  /// <returns>Disposable object that calls deferred action when disposed</returns>
```
Comment (wkirschenmann) [inline csharp/src/Deferrer.cs:71]:
```suggestion
  /// <returns>Disposable object that calls <paramref name="deferred"/> when disposed</returns>
```

File: csharp/tests/DeferTest.cs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+    var defer = Deferrer.Create(() =>
+                                {
+                                  Thread.Sleep(100);
+                                  i += 1;
+                                });
+
+    var task1 = Task.Run(() => defer.Dispose());
+    var task2 = Task.Run(() => defer.Dispose());
+
```
Comment (wkirschenmann) [inline csharp/tests/DeferTest.cs:78]: Should check here that both tasks are still running :)

File: csharp/tests/DeferTest.cs [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+  {
+    private readonly IDisposable disposable_;
+
+    public DisposableWrapper(IDisposable disposable)
+      => disposable_ = disposable;
+
+    public void Dispose()
+      => disposable_.Dispose();
+
+    ~DisposableWrapper()
```
Comment (wkirschenmann) [inline csharp/tests/DeferTest.cs:163]: Not necessary for the test. 
(nor from a functional point of view)

File: csharp/tests/DeferTest.cs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    {
+      Assert.That(i,
+                  Is.EqualTo(1));
+    }
+
+    Assert.That(i,
+                Is.EqualTo(2));
+  }
+
+  private sealed class DisposableWrapper : IDisposable
```
Comment (wkirschenmann) [inline csharp/tests/DeferTest.cs:153]: A record would be more concise.

**Review by wkirschenmann — APPROVED** [review]

**Comment by wkirschenmann** [comment]
> We might want to add AsyncDefer before merging. I'm not sure.

Can be done in another PR.

File: csharp/tests/DeferTest.cs (hunk 7 above)
Comment (lemaitre-aneo) [reply csharp/tests/DeferTest.cs:78]: It is both hard, and unnecessary. The value on `i` at the end is enough to check that there is no race condition.

**Pull request merged**